cpp/ql/src/Diagnostics/FailedExtractions.qll

```diff
@@ -0,0 +1,137 @@
+/**
+ * Provides a common hierarchy of all types of errors that can occur during extraction.
+ */
+
+import cpp
+
+/*
+ * A note about how the C/C++ extractor emits diagnostics:
+ * When the extractor frontend encounters an error, it emits a diagnostic message,
+ * that includes a message, location and severity.
+ * However, that process is best-effort and may fail (e.g. due to lack of memory).
+ * Thus, if the extractor emitted at least one diagnostic of severity discretionary
+ * error (or higher), it *also* emits a simple "There was an error during this compilation"
+ * error diagnostic, without location information.
+ * In the common case, this means that a file with one (or more) errors also gets
+ * the catch-all diagnostic.
+ * This diagnostic has the empty string as file path.
+ * We filter out these useless diagnostics if there is at least one error-level diagnostic
+ * for the affected compilation in the database.
+ * Otherwise, we show it to, to indicate that something went wrong, and we
+ * don't know what exactly happened.
+ */
+
+/**
+ * An error that, if present, leads to a file being marked as non-successfully extracted.
+ */
+class ReportableError extends Diagnostic {
+  ReportableError() {
+    (
+      this instanceof CompilerDiscretionaryError or
+      this instanceof CompilerError or
+      this instanceof CompilerCatastrophe
+    ) and
+    // Filter for the catch-all diagnostic, see note above.
+    not this.getFile().getAbsolutePath() = ""
+  }
+}
+
+private newtype TExtractionError =
+  TReportableError(ReportableError err) or
+  TCompilationFailed(Compilation c, File f) {
+    f = c.getAFileCompiled() and not c.normalTermination()
+  } or
+  // Show the catch-all diagnostic (see note above) only if we haven't seen any other error-level diagnostic
+  // for that compilation
+  TUnknownError(CompilerError err) {
+    not exists(ReportableError e | e.getCompilation() = err.getCompilation())
```

Comment:
Should we add `and err.getFile().getAbsolutePath() = ""` to ensure that this is the catch-all diagnostic and not some future diagnostic like a warning?

Reply:
That would be possible, but it is superflous:
`err` being a `CompilerError` implies that it has the severity error, not warning.
Note: This refers to the dbscheme-severity which goes from `Remark` to `Catastrophe` and is not related to the SARIF severity we're reporting in the query.
Because it is an `CompilerError` that is not a `ReportableError`, that implies `err.getFile().getAbsolutePath() = ""` by the charpred of `RepoartableError`.

If you want to future-proof this code against changes in `ReportableError` I could do that.
What do you think?

Reply:
I'd missed the type check coming from `CompilerError`. No need to change this.